### ISSUE TITLE
Pylance doesn't understand chain of nested wildcard imports

### ISSUE THREAD

**Issue by bluenote10** [microsoft/pylance-release#2200]
## Environment data

-   Language Server version: Pylance language server 2021.12.2 (pyright 211ebe44)
-   OS and version: Ubuntu 18.04
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8

## Expected behaviour

I'm demonstrating this issue based on [librosa](https://github.com/librosa/librosa). Librosa's main [`__init__.py`](https://github.com/librosa/librosa/blob/main/librosa/__init__.py) uses a standard symbol forwarding strategy. For instance, it [re-exports](https://github.com/librosa/librosa/blob/2533c0919c2d12d1497185532049e9188b248eb2/librosa/__init__.py#L229) the symbols from its `core` sub-module:

```python
from .core import *  # pylint: disable=wildcard-import
```

The `core` sub-module in turns uses a [similar approach](https://github.com/librosa/librosa/blob/main/librosa/core/__init__.py) to re-export symbols.

It looks like pylance doesn't properly understand the chain of re-exports. Even though e.g. `stft` is a function that **is** in scope of `librosa` itself (at runtime), it doesn't understand that.

![image](https://user-images.githubusercontent.com/3620703/147774399-ba3df1b5-7461-4934-81c4-6588f72c18fd.png)

Note however, that it still does see the `stft` symbol when imported through the `core` package:

![image](https://user-images.githubusercontent.com/3620703/147774478-04917bfa-5916-49c4-bc9b-8dc831174732.png)

This could indicate that the issue is caused by the chain of wildcard re-exports.

I also tried to fix the behavior by using the `__all__ = [_ for _ in dir() if not _.startswith("_")]` technique in the main `__init__.py` but it does not seem to make a difference.

## Actual behaviour

The `librosa.stft` symbol should be understood by pylance exactly the same as `librosa.core.stft`.

## Logs

(my logs didn't contain much meaningful)

```
Starting Pylance language server.
```

## Code Snippet / Additional information

```python
import librosa
import librosa.core

librosa.stft
librosa.core.stft
```

**Comment by erictraut** [comment]
The problem is that librosa is using a dynamic expression to define `__all__`. In `librosa/core/__init__.py`, you'll find the following statement:

```python
__all__ = [_ for _ in dir() if not _.startswith("_")]
```

Static type checkers like the one used in pylance don't execute the Python code; they need to be able to statically evaluate the code. Authors of Python static type checkers have agreed upon a limited set of expression forms that can be used to define `__all__`. For a fully-documented list, see [this documentation](https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface).

If I replace the above definition of `__all__` with the following supported forms, it works fine.

```python
__all__ = []
__all__.extend(convert.__all__)
__all__.extend(audio.__all__)
__all__.extend(spectrum.__all__)
__all__.extend(pitch.__all__)
__all__.extend(constantq.__all__)
__all__.extend(harmonic.__all__)
__all__.extend(fft.__all__)
__all__.extend(notation.__all__)
```

If this is a library you depend upon, I recommend contacting the library's maintainers and making them aware of this issue to see if they are willing to address it in their next update.

**Comment by bluenote10** [comment]
> The problem is that librosa is using a dynamic expression to define __all__. In librosa/core/__init__.py, you'll find the following statement:

I was indeed wondering whether these dynamic `__all__` specifications might be an issue, but why are they working then for `librosa/core/__init__.py` in the first place? The fact that they work there indicated that pylance seems to do something to resolve them, but it only doesn't work across multiple re-export stages.

**Comment by erictraut** [comment]
This doesn't have anything to do with multiple re-export stages. It will happen any time a wildcard import references a module that defines `__all__` in a way that pylance doesn't understand.

The `librosa/__init__.py` file uses the statement `from .core import *`. That's a wildcard import, so only the symbols specified in the `__all__` statement in `librosa/core/__init__.py` are included in this import. Pylance is assuming that `__all__` reference no symbols because it can't understand the dynamic statement.

I think we could handle this situation better than we currently do. If we detect an unsupported statement that manipulates `__all__`, we should fall back to assuming that all symbols are included in a wildcard statement rather than assuming none of them are.

I've implemented this change, and it "fixes" the problem you've reported, although it would still be preferable for the package author to use a form of `__all__` statement that is supported by static type checkers.

This change will be included in the next release of pylance.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202210-5-january-2022)
